TwitchChatVotingProxy/VotingReceiver/YTapiV1VotingReceiver.cs

```diff
@@ -0,0 +1,276 @@
+                return pyb64;
+            }
+            private static long[] times(int pastSec)
+            {
+                long[] ret = new long[5];
+                TimeSpan t = DateTime.UtcNow - new DateTime(1970, 1, 1);
+                long n = (long)t.TotalSeconds;
+                double ts1 = n - 1.7596119363236733;
+                double ts2 = n - 0.6937422130801;
+                double ts3 = n - pastSec + 0.5887400473806613;
+                double ts4 = n - 1876.0119363876886;
+                double ts5 = n - 0.45889037558640905;
+                ret[0] = (long)(1000000 * ts1);
+                ret[1] = (long)(1000000 * ts2);
+                ret[2] = (long)(1000000 * ts3);
+                ret[3] = (long)(1000000 * ts4);
+                ret[4] = (long)(1000000 * ts5);
+                return ret;
+            }
+
+            private string getContinuationToken(int pastSec = 5)
+            {
+                long[] timeStamps = times(pastSec);
+                return build(timeStamps);
+            }
+
+
+
+            public async Task<List<OnMessageArgs>> get()
+            {
+                if (continuation == null)
+                {
+                    continuation = getContinuationToken();
+                }
+                const string apiURL = "https://www.youtube.com/youtubei/v1/live_chat/get_live_chat?key=AIzaSyAO_FJ2SlqU8Q4STEHLGCilw_Y9_11qcW8";
+                string body = "{\"context\": {\"client\": {\"visitorData\": \"\", \"userAgent\": \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.183 Safari/537.36 Edg/86.0.622.63,gzip(gfe)\", \"clientName\": \"WEB\", \"clientVersion\": \"2.20221206.01.00\"}}, \"continuation\": \"" + continuation + "\"}";
+
+                var content = new StringContent(body);
+                var response = await client.PostAsync(apiURL, content);
+                JsonObject res = JsonObject.Parse(await response.Content.ReadAsStringAsync())!.AsObject();
+                if (res == null || res["continuationContents"] == null)
+                {
+                    // is it bad, when we constantly write this warning and try again immediatly?
+                    // or should we just throw an exception because clearly the request isn't working right
+                    logger.Warning("videoId is probably incorrect or the video is private or something else happend so that the youtube api v1 didn't return the needed data");
+                    continuation = null;
+                    return new List<OnMessageArgs>();
+                }
+                JsonObject responseData = res["continuationContents"]!["liveChatContinuation"]!.AsObject();
+                var continuations = responseData!["continuations"]!.AsArray()![0]!.AsObject();
+                JsonNode? continuationData;
+                if (continuations.TryGetPropertyValue("invalidationContinuationData", out continuationData))

```

Comment:
One shouldn't use the non-null assertion operator. Since it's technically an arbitrary JSON object, all fields could potentially be null. So you should probably use the `?` (null conditional), and check afterwords if the value exists. Maybe you could even chain the call? Like:

```cs
continuations.TryGetPropertyValue("invalidationContinuationData")
  .?["continuation]
  .?GetValue<string>();
```

But I'm not sure, since the API uses that weird out thing (which is of course not your fault lol).

Reply:
Oh, and btw. the upstream is using some other JSON parsing thing. Although, I haven't been able to merge that yet, since the c++ part keeps crashing on me :(. But maybe we could use the same library? Looks pretty clean: https://github.com/gta-chaos-mod/ChaosModV/commit/b2c9732863e8f2d6424235f55c06d13b89bf2b28#diff-19156a6e6f3132fe9e2de63a68a651114269e429e04a09898a88d5fb8af64bdcL1-R171.